internal/types/compare.go

```diff
@@ -255,3 +251,123 @@ func compareNumbers(a float64, b int64) CompareResult {
 
 	return CompareResult(bigA.Cmp(bigB))
 }
+
+// dataTypeOrderResult represents the comparison order of data types.
+type dataTypeOrderResult uint8
+
+const (
+	_ dataTypeOrderResult = iota
+	minKeyDataType
+	nullDataType
+	nanDataType
+	numbersDataType
+	stringDataType
+	objectDataType
+	arrayDataType
+	binDataType
+	objectIdDataType
+	booleanDataType
+	dateDataType
+	timestampDataType
+	regexDataType
+	maxKeyDataType
+)
+
+// defineDataType define which type has value and returns a sequence the type has.
+func defineDataType(value any) dataTypeOrderResult {
+	switch value := value.(type) {
+	case float64:
+		if math.IsNaN(value) {
+			return nanDataType
+		}
+		return numbersDataType
+	case string:
+		return stringDataType
+	case Binary:
+		return binDataType
+	case ObjectID:
+		return objectIdDataType
+	case bool:
+		return booleanDataType
+	case time.Time:
+		return dateDataType
+	case NullType:
+		return nullDataType
+	case Regex:
+		return regexDataType
+	case int32:
+		return numbersDataType
+	case Timestamp:
+		return timestampDataType
+	case int64:
+		return numbersDataType
+	}
+
+	panic(fmt.Sprintf("value cannot be defined, value is %v, data type of value is %T", value, value))
+}
+
+// numberDataTypeOrderResult represents the comparison order of numbers.
+type numberDataTypeOrderResult uint8
+
+const (
+	_ numberDataTypeOrderResult = iota
+	doubleNegativeZero
+	doubleDT
+	int32DT
+	int64DT
+)
+
+// defineNumberDataType define which number type has value and returns a sequence the type has.
+func defineNumberDataType(value any) numberDataTypeOrderResult {
+	switch value := value.(type) {
+	case float64:
+		if value == 0 && math.Signbit(value) {
+			return doubleNegativeZero
+		}
+		return doubleDT
+	case int32:
+		return int32DT
+	case int64:
+		return int64DT
+	}
+
+	panic(fmt.Sprintf("defineNumberDataType: value cannot be defined, value is %v, data type of value is %T", value, value))
+}
+
+// CompareOrder defines the data type for the two values and compares them. When the types are equal, it compares using Compare.
+func CompareOrder(a, b any, order string) CompareResult {
```

Comment:
We should use some other type for `order` instead of string, and use typed constants.

I guess we have to move `common.sortType` there and export it

Reply:
> We should use some other type for `order` instead of string...

What do you mean by that? Could you please give an example?

Reply:
Got it! Again, thank you for you time :smile:

Reply:
Should I also move all those orders of data types (`dataTypeOrderResult`, `numberDataTypeOrderResult`) to common.sortType as well?

Reply:
I think we should keep `CompareOrder` in `types` package. For that, we should move `common.sortType` to `types` package. To be able to use, we should export it.

I don't see why we need to move or export `dataTypeOrderResult` and `numberDataTypeOrderResult`

Reply:
migrated